### PR TITLE
fix: handle column order mismatches

### PR DESCRIPTION
**Pull request by pnadolny13**
Closes https://github.com/MeltanoLabs/target-snowflake/issues/116

I created a new test that creates a table then sends singer records that are in a different order. The current implementation causes the test to fail because https://github.com/MeltanoLabs/target-snowflake/actions/runs/6160685807/job/16718219357 its trying to put a string in a boolean column. After implementing Derek's suggested fix to explicitly define column names in the copy into statement, the tests pass.

cc @visch

### CODE REVIEW

File: target_snowflake/connector.py [hunk 1]
```diff
@@ -349,10 +349,11 @@ def _get_merge_from_stage_statement(
     def _get_copy_statement(self, full_table_name, schema, sync_id, file_format):
         """Get Snowflake COPY statement."""
         formatter = SnowflakeIdentifierPreparer(SnowflakeDialect())
+        formatted_properties = ", ".join([formatter.format_collation(col) for col in schema["properties"].keys()])
         column_selections = self._get_column_selections(schema, formatter)
         return (
             text(
-                f"copy into {full_table_name} from "
+                f"copy into {full_table_name} ({formatted_properties}) from "
```
Comment (visch) [inline target_snowflake/connector.py:356]: Should this be `column_selections` ? You want it to always match the snowflake table in `full_table_name` right? 

My assumption is there's column checks earlier to make seure `full_table_name` has the columns setup alredy before this runs

File: tests/core.py [hunk 2]
```diff
@@ -448,6 +448,29 @@ def validate(self) -> None:
...
+    name = "column_order_mismatch"
+
+    def setup(self) -> None:
+        connector = self.target.default_sink_class.connector_class(self.target.config)
+        table = f"{self.target.config['database']}.{self.target.config['default_target_schema']}.{self.name}".upper()
+        connector.connection.execute(
+            f"""
+            CREATE OR REPLACE TABLE {table} (
+                COL1 VARCHAR(16777216),
+                COL2 BOOLEAN
```
Comment (visch) [inline tests/core.py:464]:
```suggestion
                COL3 VARCHAR(500),
                COL4 VARCHAR(300),
                COL2 BOOLEAN
```

Comment (visch) [reply tests/core.py:464]: I think this would test my previous comment, but just did this quickly without running myself

**Review by visch — CHANGES_REQUESTED** [review]

File: target_snowflake/connector.py (hunk 1 above)
Comment (visch) [reply target_snowflake/connector.py:356]: What we did here https://github.com/MeltanoLabs/target-postgres/blob/main/target_postgres/sinks.py#L194 was just used the columns directly from the source table (the stage in this case). I wonder if that'd be even better here

Comment (pnadolny13) [reply target_snowflake/connector.py:356]: @visch `column_selection` is a formatted list used to cast to types, etc. https://github.com/MeltanoLabs/target-snowflake/blob/f560958da71d1f643428e64a70e3ae38a79b80c7/target_snowflake/connector.py#L309. This is just a list of column names.

Selecting the columns from the table before building the query is more definitive but if they didnt match then theres a bug elsewhere in the table creation or altering that happens before this step. Although couldnt that lead to us silently skipping columns if theres a bug that is failing to all the columns? Am I understanding correctly?

File: tests/core.py (hunk 2 above)
Comment (pnadolny13) [reply tests/core.py:464]: I'm not sure I follow the test case right now, can you describe in more detail?

File: target_snowflake/connector.py (hunk 1 above)
Comment (visch) [reply target_snowflake/connector.py:356]: My concern is that `column_selections` generates columns in a seperate function to `formatted_properties` right now they loop over the same schema in the same way which means order is preserved (which is the key here) but how do you keep that going in the long term so that folks don't mess that up. 

Maybe not a huge deal, but it's something that nags at me as I've seen this kind of thing blow up later as people forget the context of why things were done a certain way

File: tests/core.py (hunk 2 above)
Comment (visch) [reply tests/core.py:464]: We want to be sure that when there's extra data in the target table that the insert continues to work properly. The case this covers is if someone is iterating over the list of fields from the target table vs the list from the schema, pretty edge casey but that's where my brain went 🤷‍♂️

Comment (pnadolny13) [reply tests/core.py:466]: Makes sense, I added an unused timestamp column in there and it still passes so I think that confirms that the explicit column names are working properly.

File: target_snowflake/connector.py (hunk 1 above)
Comment (pnadolny13) [reply target_snowflake/connector.py:356]: @visch oh yeah it adds some duplicate code that could be pulled out and reused. I'll see if I can refactor quickly, otherwise I'll open a separate issue so we can squash the bug and clean up after.

Comment (visch) [reply target_snowflake/connector.py]: I like it better! Long term sqlalchmy tables / columns would probably be the slickest but that's too much lift for sure!

**Review by visch — APPROVED** [review]

**Review by edgarrmondragon — APPROVED** [review]
